mixed: prepare statements only for subcommands defined by user

muzarski opened the pull request:
# Motivation

See the discussion: https://github.com/scylladb/cql-stress/pull/86#issuecomment-2204524664.

My last PR (https://github.com/scylladb/cql-stress/pull/86) introduced a new bug - `mixed` command tries to prepare a statement for each operation (read/write/counter_read/counter_write), even though it does not create the tables.

This PR fixes it so we prepare the statements only for the operations that are going to be sampled (were specified in `ratio()` parameter by the user). We obviously then assume that the corresponding write workload was ran beforehand and that it created tables used by the `mixed` workload.

wprzytula commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/enumerated.rs:
```diff
@@ -17,6 +17,13 @@ impl<T: Copy> EnumeratedDistribution<T> {
         self.items[self.dist.sample(&mut rand::thread_rng())].0
     }
+
+    pub fn has_item(&self, t: &T) -> bool {
+        self.items
+            .iter()
+            .map(|(item, _weight)| item)
+            .any(|item| item == t)
+    }
```
Nit: a customary Rust name for such method would be `contains`.

wprzytula commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/enumerated.rs:
```diff
@@ -17,6 +17,13 @@ impl<T: Copy> EnumeratedDistribution<T> {
+    pub fn has_item(&self, t: &T) -> bool {
+        self.items
+            .iter()
+            .map(|(item, _weight)| item)
+            .any(|item| item == t)
```
Is the `map` really needed? We can do with one `any`.

muzarski replied:
It is not, good catch

wprzytula commented on src/bin/cql-stress-cassandra-stress/operation/mixed.rs:
```diff
@@ -56,10 +56,16 @@ impl OperationFactory for MixedOperationFactory {
-        let counter_write_operation = self.counter_write_operation_factory.create();
-        let read_operation = self.read_operation_factory.create();
-        let counter_read_operation = self.counter_read_operation_factory.create();
+        let write_operation = self.write_operation_factory.as_ref().map(|f| f.create());
+        let counter_write_operation = self
+            .counter_write_operation_factory
+            .as_ref()
+            .map(|f| f.create());
+        let read_operation = self.read_operation_factory.as_ref().map(|f| f.create());
+        let counter_read_operation = self
+            .counter_read_operation_factory
+            .as_ref()
+            .map(|f| f.create());
```
We could try to reduce boilerplate by introducing a generic helper function:
```rust
fn produce_operation<Factory: OperationFactory>(factory: &Option<Factory>) -> Factory::Product { 
    factory.as_ref().map(|f| f.create())
}

let write_operation = produce_operation(&self.counter_write_operation_factory);
...
```
I thought up the names; replace them with correct ones. I hope you get the idea.

wprzytula commented on src/bin/cql-stress-cassandra-stress/operation/mixed.rs:
```diff
@@ -88,16 +94,52 @@ impl MixedOperationFactory {
+            None
+        };
+        let counter_write_operation_factory = if mixed_params
+            .operation_ratio
+            .has_item(&MixedSubcommand::CounterWrite)
+        {
+            Some(CounterWriteOperationFactory::new(settings.clone(), session.clone()).await?)
+        } else {
+            None
+        };
+        let read_operation_factory = if mixed_params
+            .operation_ratio
+            .has_item(&MixedSubcommand::Read)
+        {
+            Some(
+                RegularReadOperationFactory::new(
+                    settings.clone(),
+                    session.clone(),
+                    DEFAULT_TABLE_NAME,
+                )
+                .await?,
+            )
+        } else {
+            None
+        };
+        let counter_read_operation_factory = if mixed_params
+            .operation_ratio
+            .has_item(&MixedSubcommand::CounterRead)
+        {
+            Some(
+                CounterReadOperationFactory::new(
+                    settings.clone(),
+                    session,
+                    DEFAULT_COUNTER_TABLE_NAME,
+                )
+                .await?,
+            )
+        } else {
+            None
+        };
```
WDYT about such reduction of boilerplate?
```rust
async fn conditional_create_factory<Factory: OperationFactory, Fut: Future<Output=Factory>>(kind: MixedSubcommand, f: impl FnOnce() -> Fut {
    if mixed_params.operation_ratio.has_item(&kind) {
        Some(f().await)
    } else {
        None
    }

    let read_operation_factory = conditional_crate_factory(
         MixedSubcommand::Read,
         || RegularReadOperationFactory::new(
                settings.clone(),
                session.clone(),
                DEFAULT_TABLE_NAME,
              )
    ).await;
}

fruch commented:
I've took the PR into SCT integration test, and mixed is working as expected (i.e. doesn't fail)

muzarski commented:
v2:
- addressed review comments
- moved `EnumeratedDistribution::contains` to separate impl with `PartialEq + Eq` trait bounds on `T`. Previously, we would require `T` to be `PartialEq + Eq` even for sampling methods - we obviously don't need these trait bounds to implement them.

wprzytula commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/enumerated.rs:
```diff
@@ -7,7 +7,7 @@ pub struct EnumeratedDistribution<T> {
+impl<T: Copy + PartialEq + Eq> EnumeratedDistribution<T> {
     pub fn new(items: Vec<(T, f64)>) -> Result<Self> {
         let dist = WeightedIndex::new(items.iter().map(|w| w.1))?;
 
```
I can still see `PartialEq + Eq` bounds for `new`. Is this a leftover?

muzarski commented:
v2.1: In previous comment I mentioned that we don't need `PartialEq + Eq` trait bounds to implement sampling methods. I forgot to remove them, tho...

wprzytula commented on src/bin/cql-stress-cassandra-stress/operation/mixed.rs:
```diff
@@ -111,6 +148,18 @@ impl MixedOperationFactory {
         })
     }
+
+    async fn conditional_create_factory<Factory, Fut: Future<Output = Factory>>(
+        ratios: &OperationRatio,
+        command_kind: &MixedSubcommand,
+        create_factory_fut: impl FnOnce() -> Fut,
+    ) -> Option<Factory> {
+        if ratios.contains(command_kind) {
+            Some(create_factory_fut().await)
+        } else {
+            None
+        }
+    }
```
In instantiations of this functions, is `Factory` equal to `Result<impl CassandraStressOperation>`? I struggled to find where this result comes from in outer code, but here it is. Maybe let's constrain the function further by imposing constraints on `Factory` to be `Result<impl CassandraStressOperation>`?

wprzytula commented on src/bin/cql-stress-cassandra-stress/operation/mixed.rs:
```diff
@@ -111,6 +148,18 @@ impl MixedOperationFactory {
         })
     }
+
+    async fn conditional_create_factory<Factory, Fut: Future<Output = Result<Factory>>>(
+        ratios: &OperationRatio,
+        command_kind: &MixedSubcommand,
+        create_factory_fut: impl FnOnce() -> Fut,
+    ) -> Option<Result<Factory>> {
+        if ratios.contains(command_kind) {
+            Some(create_factory_fut().await)
+        } else {
+            None
+        }
+    }
```
You still haven't constrained `Factory: CassandraStressOperationFactory`, but nvm.
